### ISSUE TITLE
False positive on Content-Type `application/activity+json`

### ISSUE THREAD

**Issue by axeleroy** [coreruleset/coreruleset#3074]
### Description

CRS blocks requests containing the `Content-Type: application/activity+json` header, even though this MIME type is legitimate: it is part of the [Activity Streams](https://www.w3.org/TR/activitystreams-core/) standard which is used by [ActivityPub](https://en.wikipedia.org/wiki/ActivityPub), itself used by many applications of the "fediverse" (think Mastodon, PeerTube, Pixelfed…)

### Audit Logs / Triggered Rule Numbers

```
[warn] 1143#1143: *17 ModSecurity: Warning. Matched "Operator `Within' with parameter `|application/x-www-form-urlencoded| |multipart/form-data| |multipart/related| |text/xml| |application/xml| |application/soap+xml| |application/x-amf| |application/json| |application/cloudevents+json|  (134 characters omitted)' against variable `TX:content_type' (Value: `|application/activity+json|' ) [file "/opt/bunkerweb/core/modsecurity/files/coreruleset/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"]
	 [line "937"]
	 [id "920420"]
	 [rev ""]
	 [msg "Request content type is not allowed by policy"]
	 [data "application/activity+json"]
	 [severity "2"]
	 [ver "OWASP_CRS/3.3.2"]
	 [maturity "0"]
	 [accuracy "0"]
	 [tag "application-multi"]
	 [tag "language-multi"]
	 [tag "platform-multi"]
	 [tag "attack-protocol"]
	 [tag "paranoia-level/1"]
	 [tag "OWASP_CRS"]
	 [tag "capec/1000/255/153"]
	 [tag "PCI/12.1"]
	 [hostname "172.X.X.X"]
	 [uri "/users/digitalfoundry/inbox"]
	 [unique_id "167266845524.629072"]
	 [ref "o0,25v268,25t:lowercase"], client: 149.X.X.X, server: birdsite.example.tld, request: "POST /users/digitalfoundry/inbox HTTP/1.1", host: "birdsite.example.tld"
[error] 1143#1143: *17 [client 149.X.X.X]
	 ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `5' ) [file "/opt/bunkerweb/core/modsecurity/files/coreruleset/rules/REQUEST-949-BLOCKING-EVALUATION.conf"]
	 [line "80"]
	 [id "949110"]
	 [rev ""]
	 [msg "Inbound Anomaly Score Exceeded (Total Score: 5)"]
	 [data ""]
	 [severity "2"]
	 [ver "OWASP_CRS/3.3.2"]
	 [maturity "0"]
	 [accuracy "0"]
	 [tag "application-multi"]
	 [tag "language-multi"]
	 [tag "platform-multi"]
	 [tag "attack-generic"]
	 [hostname "172.X.X.X"]
	 [uri "/users/digitalfoundry/inbox"]
	 [unique_id "167266845524.629072"]
	 [ref ""], client: 149.X.X.X, server: birdsite.example.tld, request: "POST /users/digitalfoundry/inbox HTTP/1.1", host: "birdsite.example.tld"
[warn] 1143#1143: *17 ModSecurity: Warning. Matched "Operator `Within' with parameter `|application/x-www-form-urlencoded| |multipart/form-data| |multipart/related| |text/xml| |application/xml| |application/soap+xml| |application/x-amf| |application/json| |application/cloudevents+json|  (134 characters omitted)' against variable `TX:content_type' (Value: `|application/activity+json|' ) [file "/opt/bunkerweb/core/modsecurity/files/coreruleset/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"]
	 [line "937"]
	 [id "920420"]
	 [rev ""]
	 [msg "Request content type is not allowed by policy"]
	 [data "application/activity+json"]
	 [severity "2"]
	 [ver "OWASP_CRS/3.3.2"]
	 [maturity "0"]
	 [accuracy "0"]
	 [tag "application-multi"]
	 [tag "language-multi"]
	 [tag "platform-multi"]
	 [tag "attack-protocol"]
	 [tag "paranoia-level/1"]
	 [tag "OWASP_CRS"]
	 [tag "capec/1000/255/153"]
	 [tag "PCI/12.1"]
	 [hostname "172.X.X.X"]
	 [uri "/users/digitalfoundry/inbox"]
	 [unique_id "167266845598.539349"]
	 [ref "o0,25v268,25t:lowercase"], client: 149.X.X.X, server: birdsite.example.tld, request: "POST /users/digitalfoundry/inbox HTTP/1.1", host: "birdsite.example.tld"
[error] 1143#1143: *17 [client 149.X.X.X]
	 ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `5' ) [file "/opt/bunkerweb/core/modsecurity/files/coreruleset/rules/REQUEST-949-BLOCKING-EVALUATION.conf"]
	 [line "80"]
	 [id "949110"]
	 [rev ""]
	 [msg "Inbound Anomaly Score Exceeded (Total Score: 5)"]
	 [data ""]
	 [severity "2"]
	 [ver "OWASP_CRS/3.3.2"]
	 [maturity "0"]
	 [accuracy "0"]
	 [tag "application-multi"]
	 [tag "language-multi"]
	 [tag "platform-multi"]
	 [tag "attack-generic"]
	 [hostname "172.X.X.X"]
	 [uri "/users/digitalfoundry/inbox"]
	 [unique_id "167266845598.539349"]
	 [ref ""], client: 149.X.X.X, server: birdsite.example.tld, request: "POST /users/digitalfoundry/inbox HTTP/1.1", host: "birdsite.example.tld"
```

### Your Environment

* CRS version: 3.3.2
* Web Server and version: [BunkerWeb 1.4.5](https://github.com/bunkerity/bunkerweb)

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by RedXanadu** [comment]
Hi @axeleroy, thanks for the report. I will take a look and get back to you soon with an answer.

**Comment by RedXanadu** [comment]
Hi @axeleroy, sorry for the delay.

The Core Rule Set only allows a very small set of "Content-Type" media types by default. The default list of allowed types can be overridden and extended in [`crs-setup.conf.example`](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/crs-setup.conf.example#L466). This makes it possible to allow additional media types for web applications that require them.

### Important:

If you're using the _default_ ModSecurity "recommended rules" (`modsecurity.conf-recommended`) then the JSON request body parser will *only* be engaged for a Content-Type media type of `application/json` **and nothing else**.

If you start to allow other media types of the form `*+json` then this could lead to a partial request body bypass. (Imagine: your WAF allows media type `foo/bar+json` but does not examine such requests using a JSON parser; your application server then parses such a request as JSON and finds/logs/executes something nasty hidden in there.)

IANA have 136(!) official JSON-related media types registered (https://www.iana.org/assignments/media-types/application.csv). For this reason, the optional 'recommended' [rule 200006](https://github.com/SpiderLabs/ModSecurity/blob/v3/master/modsecurity.conf-recommended#L32) was created, which activates the JSON body parser for any `x+json` media type. Unfortunately, this rule is commented out by default, but you may find it useful!

**Comment by axeleroy** [comment]
Thanks for your answer, I was able to update the rule to add just the MIME Type I required.
